src/main/java/software/amazon/awssdk/s3accessgrants/cache/S3AccessGrantsCachedBucketRegionResolver.java

```diff
@@ -0,0 +1,170 @@
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.BUCKET_REGION_EXPIRE_CACHE_AFTER_WRITE_SECONDS;
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.MAX_BUCKET_REGION_EXPIRE_CACHE_AFTER_WRITE_SECONDS;
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.MAX_BUCKET_REGION_CACHE_SIZE;
+
+public class S3AccessGrantsCachedBucketRegionResolver implements S3AccessGrantsBucketRegionResolver{
+
+
+    private int maxCacheSize;
+
+    private int expireCacheAfterWriteSeconds;
+
+    private static Cache<String, Region> cache;
+
+    public int getMaxCacheSize() {
+        return maxCacheSize;
+    }
+
+
+    public int getExpireCacheAfterWriteSeconds() {
+        return expireCacheAfterWriteSeconds;
+    }
+
+    private S3Client s3Client;
+
+    private static final Logger logger = Logger.loggerFor(S3AccessGrantsCachedBucketRegionResolver.class);
+
+    public int maxCacheSize() {
+        return maxCacheSize;
+    }
+
+    public int expireCacheAfterWriteSeconds() {
+        return expireCacheAfterWriteSeconds;
+    }
+
+    protected CacheStats getCacheStats() { return cache.stats(); }
+
+    S3AccessGrantsCachedBucketRegionResolver() {
+        this.maxCacheSize = BUCKET_REGION_CACHE_SIZE;
+        this.expireCacheAfterWriteSeconds = BUCKET_REGION_EXPIRE_CACHE_AFTER_WRITE_SECONDS;
+    }
+
+
+    public S3AccessGrantsCachedBucketRegionResolver.Builder toBuilder() {
+        return new S3AccessGrantsCachedBucketRegionResolver.BuilderImpl(this);
+    }
+
+    public static S3AccessGrantsCachedBucketRegionResolver.Builder builder() {
+        return new S3AccessGrantsCachedBucketRegionResolver.BuilderImpl();
+    }
+
+    @Override
+    public Region resolve(String bucket, S3Client s3Client) throws S3Exception {
+
+        this.s3Client = s3Client;
+        Region bucketRegion = cache.getIfPresent(bucket);
+        if(bucketRegion == null) {
+            logger.debug(() -> "bucket region not available in cache, fetching the region from the service!");
+            if (s3Client == null) {
+                throw new IllegalArgumentException("S3Client is required for the bucket region resolver!");
+            }
+            bucketRegion = resolveFromService(bucket);
+            if(bucketRegion != null) {
+                cache.put(bucket, bucketRegion);
+            }
+        } else {
+            logger.debug(() -> "bucket region available in cache!");
+        }
+        return bucketRegion;
+
+    }
+
+    private Region resolveFromService(String bucket) {
+        String resolvedRegion = null;
+        try {
+                logger.info(() -> "making a call to S3 for determining the bucket region!");
+                HeadBucketRequest bucketLocationRequest = HeadBucketRequest.builder().bucket(bucket).build();
+                HeadBucketResponse headBucketResponse = s3Client.headBucket(bucketLocationRequest);
```

Comment:
This does not seem thread safe. The s3Client you have here could be replaced with another s3Client when you call public resolve method above which replaces the local s3Client.

Reply:
Yes, this does cause confusion of s3Client being replaced. Actually it is one single instance being passed around. I was trying to follow the same pattern of other cache layers. However, thinking about it, it would be a good idea to just have one S3Client configured at the time of initializing this cache. That way the bucket regions cache will have only one copy of S3Client to use.